File: 단어 변환/단어 변환.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;

        if (!Arrays.asList(words).contains(target))
            return answer;

        Queue<String> queue = new LinkedList<>();
        queue.add(begin);

        HashMap<String, Boolean> map = new HashMap<>();
        for (String word : words)
            map.put(word, false);

        map.replace(begin, true);

        while (queue.size() > 0) {
            int l = queue.size();
            for (int i=0; i<l; i++) {
                begin = queue.poll();
                for (String word : words) {
                    if (!map.get(word)) {
                        int dup = 0;
                        for (int j=0; j<word.length(); j++) {
                            if (!begin.split("")[j].equals(word.split("")[j])) dup++;
                        }
                        if (dup == 1) {
                            if (word.equals(target)) return ++answer;
                            map.replace(word, true);
                            queue.add(word);
                        }
                    }
                }
            }
            answer++;
        }
        return 0;
    }
}

public class Main {
    public static void main(String[] args) {
        Solution sol = new Solution();

        String begin = "hit";
        String target = "cog";
        String[] words = {"hot", "dot", "dog", "lot", "log", "cog"};

        System.out.println(sol.solution(begin, target, words));
        // 4
    }
}

File: 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = 0
    word_len = len(words[0])

    if target not in set(words):
        return 0

    checked = {word: False for word in words}

    BFS = deque([begin])

    while BFS:
        l = len(BFS)
        for i in range(l):
            begin = BFS.popleft()
            for word in words:  # 모든 단어들에 대해서
                if not checked[word]:   # 한번 검사한 적 업는 단어일 때
                    tmp = 0
                    for idx in range(word_len):
                        if begin[idx] != word[idx]: tmp += 1    # 중복도 검사
                    if tmp == 1:    # 하나의 철자만 다르다면
                        if word == target: return answer + 1    # 그 단어가 목표 단어일 때 결과 반환

                        checked[word] = True    # 검사한 단어 리스트에 추가
                        BFS.append(word)    # 검사해야할 리스트에 추가
        answer += 1
    return 0


print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"])) # 4
